design/open/staging-compaction.md

```diff
@@ -0,0 +1,172 @@
+# Proposal: Staging Compaction
+
+## Problem Description
+
+lakeFS staging area is built on top of the [KV store](../accepted/metadata_kv/index.md).
+In short, the structure for the `Branch` entity in the kv-store is as follows:
+```go
+type Branch struct {
+	CommitID     CommitID
+	StagingToken StagingToken
+	// SealedTokens - Staging tokens are appended to the front, this allows building the diff iterator easily
+	SealedTokens []StagingToken
+}
+```
+
+Uncommitted entries are read from the staging token first, then from the 
+sealed tokens by order. Writes are performed on the staging token.
+
+The KV design has proven to meet lakeFS requirements for its consistency 
+guarantees. For most use-cases of uncommitted areas in lakeFS, it does that 
+efficiently. However, there are some cases where it falls short, for example 
+this [issue](https://github.com/treeverse/lakeFS/issues/2092).  
+There might be other cases where the structure of the `Branch` entity impacts 
+the performance of reading from the staging area, for example when the 
+number of sealed tokens is large (N)[^1] and reading a missing entry requires 
+reading from all N sealed tokens. 
+
+## Goals
+
+- Preserve lakeFS consistency guarantees for all scenarios.
+- A more efficient way to read from branches with a large number of 
+  tombstones (i.e. Fix the issue mentioned above).
+- Preserve lakeFS performance for other scenarios.[^2]
```

Comment:
I think that "other scenarios" can cause also be accessing the blockstore. Iterate over metarange can cause slowdown while we requrire to download blocks (unless they are cached). The currently iterator uses next for both uncommitted and committed so I assume we can say that we have the same tombstone issue while we scan over committed data if the staging/uncommtited entries are full of changes too and we "pay" in performance while scan data that we will served but still scan it.

Reply:
I don't think it's true. The `CompactedMetaRange` will not contain just the uncommitted entries. It will contain a snapshot of the branch at the time of the compaction, i.e. as if commit ran at the same time. Meaning, tombstones will not be part of the `CompactedMetaRange` as we do not store deletions in a commit. So we can't have the same tombstone issue.
However, you're pointing on a different potential problem. We're replacing load from the kv-store with load on the object store.